### ISSUE TITLE
Kubernetes pod/containers monitoring

### ISSUE THREAD

**Issue by ilyam8** [netdata/netdata#5387]
##### Feature idea summary

**Separate plugin for kubernetes pod/containers monitoring.**

Currently [cgroups](https://github.com/netdata/netdata/tree/master/collectors/cgroups.plugin) plugin is responsible for monitoring kubernetes pod/containers monitoring. It works, but no as good as we would like.

We have 2 options:
 - modify cgroups plugin
 - write a new plugin, which uses kubelet stats API to retrieve info about running pods/containers

2nd options seems more natural. 

**Bad thing about kubelet stats API**:  i may be wrong, but according my tests (minikube), at least by default metrics from it are not _realtime_. Need to investigate this. @cakrit 

A lot of good ideas/suggestions in https://github.com/netdata/netdata/issues/3520

___
Pod info from kubelet stats API

</br>
<details><summary>pod stats</summary>
<p>

```json
{
    "podRef": {
        "name": "nginx-5c7588df-bn6br",
        "namespace": "default",
        "uid": "ca1a29f5-2e05-11e9-91f9-0800271a908f"
    },
    "startTime": "2019-02-11T14:03:21Z",
    "containers": [
        {
            "name": "nginx",
            "startTime": "2019-02-11T14:03:24Z",
            "cpu": {
                "time": "2019-02-11T19:16:50Z",
                "usageNanoCores": 0,
                "usageCoreNanoSeconds": 20713304
            },
            "memory": {
                "time": "2019-02-11T19:16:50Z",
                "usageBytes": 2375680,
                "workingSetBytes": 2314240,
                "rssBytes": 1409024,
                "pageFaults": 1155,
                "majorPageFaults": 19
            },
            "rootfs": {
                "time": "2019-02-11T19:16:50Z",
                "availableBytes": 13237149696,
                "capacityBytes": 17293533184,
                "usedBytes": 65536,
                "inodesFree": 9658751,
                "inodes": 9732096,
                "inodesUsed": 14
            },
            "logs": {
                "time": "2019-02-11T19:16:50Z",
                "availableBytes": 13237149696,
                "capacityBytes": 17293533184,
                "usedBytes": 28672,
                "inodesFree": 9658751,
                "inodes": 9732096,
                "inodesUsed": 73345
            },
            "userDefinedMetrics": null
        }
    ],
    "cpu": {
        "time": "2019-02-11T19:16:38Z",
        "usageNanoCores": 0,
        "usageCoreNanoSeconds": 40964297
    },
    "memory": {
        "time": "2019-02-11T19:16:38Z",
        "usageBytes": 2789376,
        "workingSetBytes": 2727936,
        "rssBytes": 1449984,
        "pageFaults": 0,
        "majorPageFaults": 0
    },
    "network": {
        "time": "2019-02-11T19:16:40Z",
        "name": "eth0",
        "rxBytes": 4594,
        "rxErrors": 0,
        "txBytes": 2652,
        "txErrors": 0,
        "interfaces": [
            {
                "name": "eth0",
                "rxBytes": 4594,
                "rxErrors": 0,
                "txBytes": 2652,
                "txErrors": 0
            },
            {
                "name": "sit0",
                "rxBytes": 0,
                "rxErrors": 0,
                "txBytes": 0,
                "txErrors": 0
            }
        ]
    },
    "volume": [
        {
            "time": "2019-02-11T14:04:14Z",
            "availableBytes": 1043763200,
            "capacityBytes": 1043775488,
            "usedBytes": 12288,
            "inodesFree": 254819,
            "inodes": 254828,
            "inodesUsed": 9,
            "name": "default-token-gk662"
        }
    ],
    "ephemeral-storage": {
        "time": "2019-02-11T19:16:50Z",
        "availableBytes": 13237149696,
        "capacityBytes": 17293533184,
        "usedBytes": 94208,
        "inodesFree": 9658751,
        "inodes": 9732096,
        "inodesUsed": 14
    }
}
```

</p>
</details>

___ 

We plan to collect (striked out means it won't be collected):

**Summary pod metrics**:
  * cpu
  * memory
  * netowrk
    * defaut ingerface metrics (?)
    * ~~per iface~~
  * volume
    * per volume
  * ~~ephemeral-storage~~

**Per container metrics**:
  * cpu
  * memory
  * ~~logfs~~
  * ~~rootfs~~

___
**Representation**

This is the tough question. There is no way to represent it correctly (hierarchy) with the current dashboard.

We will use this [approach](https://github.com/netdata/netdata/issues/3520#issuecomment-373094900)

`Kubernetes` -> `POD Name` and `POD Name + Container Name` -> `graphs`

So it will look something like this:
- Kubernetes
  - example-app-5b8d7c86f8-ht678
    - cpu
    - mem
  - example-app-5b8d7c86f8-ht678 - app
    - cpu
    - mem
  - example-app-5b8d7c86f8-ht678 - web
    - cpu
    - mem
  - example-redis-1234
    - cpu
    - mem
  - example-redis-1234 - twemproxy
    - cpu
    - mem
  - example-redis-1234 - redis_6389
    - cpu
    - mem
##### Expected behavior

new plugin for monitoring kubernetes pods/containers

@geekhead @shadycuz

**Comment by ilyam8** [comment]
I think i am wrong about `network`, i have no idea what is `default` interface.

Example:

<details><summary>see screenshot</summary>
<p>

![screenshot_20190214_114346](https://user-images.githubusercontent.com/22274335/52774429-32d65380-304e-11e9-89dd-429fcec2209c.png)
</p>
</details>

</br>

It looks like we need to collect data for all nics (except that sit0), not only for `default`.

**Comment by ilyam8** [comment]
Ok, kubelet `stats/summary` endpoint always returns cached data and this is hardcoded.

https://github.com/kubernetes/kubernetes/blob/c90bf8d8ea6e02014008ee9e1afe490bee4a0bcf/pkg/kubelet/server/stats/handler.go#L201-L215

On my local install cached data updates every ~8-15 seconds.

This is a big blocker.

**Comment by cakrit** [comment]
As discussed this morning, we will be looking into doing exactly what `cadvisor` does instead, without its performance limitations. The implementation will need to be in c, extending what we are already collecting. We'll be the only ones able to offer per second monitoring for kubernetes, and we can eventually be providers of [resource metrics](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/resource-metrics-api.md) and (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/custom-metrics-api.md).

**Comment by ilyam8** [comment]
Ok, the problem with it that pods/containers monitoring can't be done only collecting info from cgroups.

We need:
 * metrics
 * labels associated with these metrics
 * dynamic charts, should be created on demand (query)

Example: cubernetes cluster with several projects
 - ns/project-a
 - ns/project-b
 - ns/project-c

Project representation:

![screenshot_20190225_113049](https://user-images.githubusercontent.com/22274335/53324062-e6fa9880-38f0-11e9-8418-49d2ee3891c4.png)

So we need to be able to chart following charts:
 - per project (per namespace)
 - per pod group
 - per pod
 - per container

That is drill down approach. You don't want to dig into container metrics from start, it makes no sense. 

This is only example, there can be more layers. It depends on project architecture.

And the aggregation is only achievable using labels.

**Comment by shadycuz** [comment]
This makes a lot of sense. It might even be helpful to look at some other monitoring tools. I can give you access to my grafana/prometheus setup if you want.

**Comment by cakrit** [comment]
We went with the modification of the `cgroups` plugin as a first step, which was the original purpose of this issue. So this onew will be closed by #5576 and we will create a different issue for the organization/presentation of the information, which will be much more difficult. We also have a beta helm chart now, see #3727. We'll be organizing the TODOs in a master ticket in the next couple of days.
